src/accountant.rs

```diff
@@ -76,28 +77,28 @@ impl Accountant {
             return Err(AccountingError::InvalidTransferSignature);
         }
 
-        if let Some(x) = self.balances.get_mut(&tr.from) {
+        if let Some(x) = self.balances.write().unwrap().get_mut(&tr.from) {
             *x -= tr.tokens;
         }
 
         let mut plan = tr.plan.clone();
-        plan.apply_witness(&Witness::Timestamp(self.last_time));
+        plan.apply_witness(&Witness::Timestamp(*self.last_time.read().unwrap()));
 
         if let Some(ref payment) = plan.final_payment() {
-            apply_payment(&mut self.balances, payment);
+            apply_payment(&mut self.balances.write().unwrap(), payment);
```

Comment:
will it try to get write lock for balances twice?

Reply:
No, a subtlety of Rust, by not putting `balances` into variable, it's only in scope while the righthand side of that expression is evaluated. If I put the first instance into a variable, then the second would immediately cause a panic.